lock-api/src/main/java/com/palantir/lock/client/PartitioningTimelockService.java

```diff
@@ -0,0 +1,176 @@
+        this(Clock.systemUTC(), timelockService);
+    }
+
+    @Override
+    public TimelockService delegate() {
+        return timelockService;
+    }
+
+    private Set<LockToken> addProxies(Set<LockToken> maybeProxies) {
+        if (proxyLockTokens.isEmpty()) {
+            return maybeProxies;
+        } else {
+            return maybeProxies.stream()
+                    .flatMap(token ->
+                            Streams.concat(
+                                    Stream.of(token),
+                                    Optional.ofNullable(proxyLockTokens.remove(token))
+                                            .orElse(Collections.emptySet()).stream()))
+                    .collect(toSet());
+        }
+    }
+
+    @Override
+    public Set<LockToken> unlock(Set<LockToken> unenrichedTokens) {
+        Set<LockToken> tokens = addProxies(unenrichedTokens);
+        if (tokens.size() <= UNLOCK_BATCH_THRESHOLD) {
+            return timelockService.unlock(tokens);
+        }
+        return Streams.stream(Iterables.partition(tokens, UNLOCK_BATCH_THRESHOLD))
+                .map(ImmutableSet::copyOf)
+                .map(timelockService::unlock)
+                .flatMap(Collection::stream)
+                .collect(toSet());
+    }
+
+    @Override
+    public void tryUnlock(Set<LockToken> unenrichedTokens) {
+        Set<LockToken> tokens = addProxies(unenrichedTokens);
+        Streams.stream(Iterables.partition(tokens, UNLOCK_BATCH_THRESHOLD))
+                .map(ImmutableSet::copyOf)
+                .forEach(timelockService::tryUnlock);
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        if (request.getLockDescriptors().size() <= LOCK_BATCH_THRESHOLD) {
+            return timelockService.lock(request);
+        }
+
+        Instant startTime = clock.instant();
+        Duration initialAcquireTimeout = Duration.ofMillis(request.getAcquireTimeoutMs());
+        List<LockDescriptor> sortedLockDescriptors =
+                Ordering.natural().immutableSortedCopy(request.getLockDescriptors());
+        BatchedLockDescriptors descriptors = new BatchedLockDescriptors();
+        try {
+            for (List<LockDescriptor> batch : Iterables.partition(sortedLockDescriptors, LOCK_BATCH_THRESHOLD)) {
+                Optional<Duration> acquireTimeout = acquireTimeout(startTime, initialAcquireTimeout);
+                if (!acquireTimeout.isPresent()) {
+                    return descriptors.failure();
+                }
+
+                LockRequest batchedRequest = ImmutableLockRequest.builder()
+                        .from(request)
+                        .acquireTimeoutMs(acquireTimeout.get().toMillis())
+                        .lockDescriptors(batch)
+                        .build();
+                LockResponse response = timelockService.lock(batchedRequest);
+                if (!response.wasSuccessful()) {
+                    return descriptors.failure();
```

Comment:
propagate the lock response?

Reply:
ie copy the failure mode

Reply:
yep - there is no other failure mode than 'timed out'.